<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Transitional//EN"
  "http://www.w3.org/TR/xhtml1/DTD/xhtml1-transitional.dtd">


<html xmlns="http://www.w3.org/1999/xhtml">
  <head>
    <meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
    
    <title>Quick start tutorial &#8212; ADENINE 0.1.4 documentation</title>
    
    <link rel="stylesheet" href="_static/slipGuru.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    
    <script type="text/javascript">
      var DOCUMENTATION_OPTIONS = {
        URL_ROOT:    './',
        VERSION:     '0.1.4',
        COLLAPSE_INDEX: false,
        FILE_SUFFIX: '.html',
        HAS_SOURCE:  true
      };
    </script>
    <script type="text/javascript" src="_static/jquery.js"></script>
    <script type="text/javascript" src="_static/underscore.js"></script>
    <script type="text/javascript" src="_static/doctools.js"></script>
    <script type="text/javascript" src="https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>
    <link rel="top" title="ADENINE 0.1.4 documentation" href="index.html" />
    <link rel="prev" title="Adenine (A Data ExploratioN pIpeliNE)" href="index.html" />
<meta name="keywords"
      content="SlipGuru, 'University of Genoa', statistical learning,
      computational biology, Python, distributed, computing, parallel" />
<map id="logosmap" name="logosmap">
    <area shape="rect" alt="SlipGURU" title="SlipGURU" coords="0,0,89,112"
          href="http://slipguru.disi.unige.it/" />
    <area shape="rect" alt="Dipartimento di Informatica e Scienze dell'Informazione"
          title="Dipartimento di Informatica e Scienze dell'Informazione"
          coords="95,4,200,34" href="http://www.disi.unige.it/" />
    <area shape="rect" alt="Università Degli Studi di Genova"
          title="Università Degli Studi di Genova" coords="124,48,171,107"
          href="http://www.unige.it/" />
</map>

  </head>
  <body role="document">
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Adenine (A Data ExploratioN pIpeliNE)"
             accesskey="P">previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ADENINE 0.1.4 documentation</a> &#187;</li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="quick-start-tutorial">
<span id="tutorial"></span><h1>Quick start tutorial<a class="headerlink" href="#quick-start-tutorial" title="Permalink to this headline">¶</a></h1>
<p><strong>Adenine*</strong> can be installed using standard Python tools (with
administrative or sudo permissions on GNU-Linux platforms):</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ pip install adenine
</pre></div>
</div>
<div class="section" id="installation-from-sources">
<h2>Installation from sources<a class="headerlink" href="#installation-from-sources" title="Permalink to this headline">¶</a></h2>
<p>If you like to manually install <strong>Adenine</strong>, download the .zip or .tar.gz archive
from <a class="reference external" href="http://slipguru.github.io/adenine/">http://slipguru.github.io/adenine/</a>. Then extract it and move into the root directory:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ unzip slipguru-adenine-|release|.zip
$ cd adenine-|release|/
</pre></div>
</div>
<p>or:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ tar xvf slipguru-adenine-|release|.tar.gz
$ cd adenine-|release|/
</pre></div>
</div>
<p>Otherwise you can clone our <a class="reference external" href="https://github.com/slipguru/adenine">GitHub repository</a>:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ git clone https://github.com/slipguru/adenine.git
</pre></div>
</div>
<p>From here, you can follow the standard Python installation step:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ python setup.py install
</pre></div>
</div>
<p>After <strong>Adenine</strong> installation, you should have access to two scripts,
named with a common <code class="docutils literal"><span class="pre">ade_</span></code> prefix:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ade_&lt;TAB&gt;
ade_analysis.py    ade_run.py
</pre></div>
</div>
<p>This tutorial assumes that you downloaded and extracted <strong>Adenine</strong>
source package which contains a <code class="docutils literal"><span class="pre">examples\data</span></code> directory with some data files (<code class="docutils literal"><span class="pre">.npy</span></code> or <code class="docutils literal"><span class="pre">.csv</span></code>) which will be used to show <strong>Adenine</strong> functionalities.</p>
<p><strong>Adenine</strong> needs only 3 ingredients:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">n_samples</span> <span class="pre">x</span> <span class="pre">n_variables</span></code> input matrix</li>
<li><code class="docutils literal"><span class="pre">n_samples</span> <span class="pre">x</span> <span class="pre">1</span></code> output vector (optional)</li>
<li><code class="docutils literal"><span class="pre">configuration</span></code> file</li>
</ul>
</div>
<div class="section" id="input-data-format">
<h2>Input data format<a class="headerlink" href="#input-data-format" title="Permalink to this headline">¶</a></h2>
<p>Input data are assumed to be:</p>
<ul class="simple">
<li><code class="docutils literal"><span class="pre">numpy</span></code> array stored in <code class="docutils literal"><span class="pre">.npy</span></code> files organized with a row for each sample and a column for each feature,</li>
<li>tabular data stored in comma separated <code class="docutils literal"><span class="pre">.csv</span></code> files presenting the variables header on the first row and the sample indexes on the first column,</li>
<li>toy examples available from <code class="docutils literal"><span class="pre">adenine.utils.data_source</span></code> function.</li>
</ul>
</div>
<div class="section" id="configuration-file">
<span id="configuration"></span><h2>Configuration File<a class="headerlink" href="#configuration-file" title="Permalink to this headline">¶</a></h2>
<p><strong>Adenine</strong> configuration file is a standard Python script. It is
imported as a module, then all the code is executed. In this file the user can define all the option needed to read the data and to create the pipelines.</p>
<div class="highlight-python"><div class="highlight"><pre><span></span><span class="ch">#!/usr/bin/env python</span>
<span class="c1"># -*- coding: utf-8 -*-</span>
<span class="sd">&quot;&quot;&quot;Configuration file for adenine.&quot;&quot;&quot;</span>

<span class="kn">from</span> <span class="nn">adenine.utils</span> <span class="kn">import</span> <span class="n">data_source</span>

<span class="c1"># --------------------------  EXPERMIENT INFO ------------------------- #</span>
<span class="n">exp_tag</span> <span class="o">=</span> <span class="s1">&#39;_experiment&#39;</span>
<span class="n">output_root_folder</span> <span class="o">=</span> <span class="s1">&#39;results&#39;</span>
<span class="n">plotting_context</span> <span class="o">=</span> <span class="s1">&#39;notebook&#39;</span>  <span class="c1"># one of {paper, notebook, talk, poster}</span>
<span class="n">file_format</span> <span class="o">=</span> <span class="s1">&#39;pdf&#39;</span>  <span class="c1"># or &#39;png&#39;</span>

<span class="c1"># ----------------------------  INPUT DATA ---------------------------- #</span>
<span class="c1"># Load an example dataset or specify your input data in tabular format</span>
<span class="n">data_file</span> <span class="o">=</span> <span class="s1">&#39;data.csv&#39;</span>
<span class="n">labels_file</span> <span class="o">=</span> <span class="s1">&#39;labels.csv&#39;</span>  <span class="c1"># OPTIONAL</span>
<span class="n">samples_on</span> <span class="o">=</span> <span class="s1">&#39;rows&#39;</span>  <span class="c1"># if samples lie on columns use &#39;cols&#39; or &#39;col&#39;</span>
<span class="n">data_sep</span> <span class="o">=</span> <span class="s1">&#39;,&#39;</span>  <span class="c1"># the data separator. e.g., &#39;,&#39;, &#39;\t&#39;, &#39; &#39;, ...</span>
<span class="n">X</span><span class="p">,</span> <span class="n">y</span><span class="p">,</span> <span class="n">feat_names</span><span class="p">,</span> <span class="n">index</span> <span class="o">=</span> <span class="n">data_source</span><span class="o">.</span><span class="n">load</span><span class="p">(</span><span class="s1">&#39;custom&#39;</span><span class="p">,</span>
                                           <span class="n">data_file</span><span class="p">,</span> <span class="n">labels_file</span><span class="p">,</span>
                                           <span class="n">samples_on</span><span class="o">=</span><span class="n">samples_on</span><span class="p">,</span>
                                           <span class="n">sep</span><span class="o">=</span><span class="n">data_sep</span><span class="p">)</span>

<span class="c1"># -----------------------  PIPELINES DEFINITION ------------------------ #</span>
<span class="c1"># --- Missing values imputing --- #</span>
<span class="n">step0</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;Impute&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;missing_values&#39;</span><span class="p">:</span> <span class="s1">&#39;NaN&#39;</span><span class="p">,</span>
                            <span class="s1">&#39;strategy&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;median&#39;</span><span class="p">,</span>
                                         <span class="s1">&#39;mean&#39;</span><span class="p">,</span>
                                         <span class="s1">&#39;nearest_neighbors&#39;</span><span class="p">]}]}</span>

<span class="c1"># --- Data preprocessing --- #</span>
<span class="n">step1</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;None&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">],</span> <span class="s1">&#39;Recenter&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">],</span> <span class="s1">&#39;Standardize&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">],</span>
         <span class="s1">&#39;Normalize&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;norm&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;l1&#39;</span><span class="p">,</span> <span class="s1">&#39;l2&#39;</span><span class="p">]}],</span>
         <span class="s1">&#39;MinMax&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;feature_range&#39;</span><span class="p">:</span> <span class="p">[(</span><span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">),</span> <span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="mi">1</span><span class="p">)]}]}</span>

<span class="c1"># --- Unsupervised features learning --- #</span>
<span class="c1"># affinity ca be precumputed for SE</span>
<span class="n">step2</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;PCA&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;n_components&#39;</span><span class="p">:</span> <span class="mi">3</span><span class="p">}],</span>
         <span class="s1">&#39;IncrementalPCA&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">],</span>
         <span class="s1">&#39;RandomizedPCA&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">],</span>
         <span class="s1">&#39;KernelPCA&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;kernel&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;linear&#39;</span><span class="p">,</span> <span class="s1">&#39;rbf&#39;</span><span class="p">,</span> <span class="s1">&#39;poly&#39;</span><span class="p">]}],</span>
         <span class="s1">&#39;Isomap&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;n_neighbors&#39;</span><span class="p">:</span> <span class="mi">5</span><span class="p">}],</span>
         <span class="s1">&#39;LLE&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;n_neighbors&#39;</span><span class="p">:</span> <span class="mi">5</span><span class="p">,</span>
                         <span class="s1">&#39;method&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;standard&#39;</span><span class="p">,</span> <span class="s1">&#39;modified&#39;</span><span class="p">,</span>
                                    <span class="s1">&#39;hessian&#39;</span><span class="p">,</span> <span class="s1">&#39;ltsa&#39;</span><span class="p">]}],</span>
         <span class="s1">&#39;SE&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;affinity&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;nearest_neighbors&#39;</span><span class="p">,</span> <span class="s1">&#39;rbf&#39;</span><span class="p">]}],</span>
         <span class="s1">&#39;MDS&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;metric&#39;</span><span class="p">:</span> <span class="bp">True</span><span class="p">}],</span>
         <span class="s1">&#39;tSNE&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">],</span>
         <span class="s1">&#39;RBM&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;n_components&#39;</span><span class="p">:</span> <span class="mi">256</span><span class="p">}],</span>
         <span class="s1">&#39;None&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">]</span>
         <span class="p">}</span>

<span class="c1"># --- Clustering --- #</span>
<span class="c1"># affinity ca be precumputed for AP, Spectral and Hierarchical</span>
<span class="n">step3</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;KMeans&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;n_clusters&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="s1">&#39;auto&#39;</span><span class="p">]}],</span>
         <span class="s1">&#39;AP&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;preference&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;auto&#39;</span><span class="p">]}],</span>
         <span class="s1">&#39;MS&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">],</span>
         <span class="s1">&#39;Spectral&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;n_clusters&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">]}],</span>
         <span class="s1">&#39;Hierarchical&#39;</span><span class="p">:</span> <span class="p">[</span><span class="bp">False</span><span class="p">,</span> <span class="p">{</span><span class="s1">&#39;n_clusters&#39;</span><span class="p">:</span> <span class="p">[</span><span class="mi">3</span><span class="p">,</span> <span class="mi">8</span><span class="p">],</span>
                                  <span class="s1">&#39;affinity&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;manhattan&#39;</span><span class="p">,</span> <span class="s1">&#39;euclidean&#39;</span><span class="p">],</span>
                                  <span class="s1">&#39;linkage&#39;</span><span class="p">:</span>  <span class="p">[</span><span class="s1">&#39;ward&#39;</span><span class="p">,</span> <span class="s1">&#39;complete&#39;</span><span class="p">,</span> <span class="s1">&#39;average&#39;</span><span class="p">]}]</span>
         <span class="p">}</span>
</pre></div>
</div>
</div>
<div class="section" id="experiment-runner">
<span id="experiment"></span><h2>Experiment runner<a class="headerlink" href="#experiment-runner" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">ade_run.py</span></code> script, executes the full <strong>Adenine</strong> framework. The prototype is the following:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ade_run.py ade_config.py
</pre></div>
</div>
<p>When launched, the script reads the data, then it creates and runs each pipeline saving the results in a tree-like structure which has the current folder as root.</p>
</div>
<div class="section" id="results-analysis">
<span id="analysis"></span><h2>Results analysis<a class="headerlink" href="#results-analysis" title="Permalink to this headline">¶</a></h2>
<p>The <code class="docutils literal"><span class="pre">ade_analysis.py</span></code> script provides useful summaries and graphs from the results of the experiment. This script accepts as only parameter a result directory
already created:</p>
<div class="highlight-default"><div class="highlight"><pre><span></span>$ ade_analysis.py result-dir
</pre></div>
</div>
<p>The script produces a set of textual and graphical results. An output example obtained by one of the implemented pipelines is represented below.</p>
<a class="reference internal image-reference" href="_images/pca.png"><img alt="broken link" src="_images/pca.png" style="width: 400.0px; height: 400.0px;" /></a>
<a class="reference internal image-reference" href="_images/kpca.png"><img alt="broken link" src="_images/kpca.png" style="width: 400.0px; height: 400.0px;" /></a>
<p>You can reproduce the example above specifying <code class="docutils literal"><span class="pre">data_source.load('circles')</span></code> in the configuration file.</p>
</div>
<div class="section" id="example-dataset">
<h2>Example dataset<a class="headerlink" href="#example-dataset" title="Permalink to this headline">¶</a></h2>
<p>An example dataset can be dowloaded <code class="xref download docutils literal"><span class="pre">here</span></code>. The dataset is a random extraction of 801 samples (with dimension 20531) measuring RNA-Seq gene expression of patients affected by 5 different types of tumor: breast invasive carcinoma (BRCA), kidney renal clear cell carcinoma (KIRC), colon  (COAD), lung  (LUAD) and prostate adenocarcinoma (PRAD). The full dataset is maintained by The Cancer Genome Atlas Pan-Cancer Project [1] and we refer to the <a class="reference external" href="https://www.synapse.org/#!Synapse:syn4301332">original repository</a> for furher details.</p>
</div>
<div class="section" id="reference">
<h2>Reference<a class="headerlink" href="#reference" title="Permalink to this headline">¶</a></h2>
<p>[1] Weinstein, John N., et al. &#8220;The cancer genome atlas pan-cancer analysis project.&#8221; Nature genetics 45.10 (2013): 1113-1120.</p>
</div>
</div>


          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
            <p class="logo"><a href="index.html">
              <img class="logo" src="_static/adenine_logo.png" alt="Logo"/>
            </a></p>
   
      
  <h3><a href="index.html">Table Of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">Quick start tutorial</a><ul>
<li><a class="reference internal" href="#installation-from-sources">Installation from sources</a></li>
<li><a class="reference internal" href="#input-data-format">Input data format</a></li>
<li><a class="reference internal" href="#configuration-file">Configuration File</a></li>
<li><a class="reference internal" href="#experiment-runner">Experiment runner</a></li>
<li><a class="reference internal" href="#results-analysis">Results analysis</a></li>
<li><a class="reference internal" href="#example-dataset">Example dataset</a></li>
<li><a class="reference internal" href="#reference">Reference</a></li>
</ul>
</li>
</ul>

   

   
      
  <h4>Previous topic</h4>
  <p class="topless"><a href="index.html"
                        title="previous chapter">Adenine (A Data ExploratioN pIpeliNE)</a></p>
   

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/tutorial.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
    
<div id="searchbox" style="display: none" role="search">
  <h3>Quick search</h3>
    <form class="search" action="search.html" method="get">
      <div><input type="text" name="q" /></div>
      <div><input type="submit" value="Go" /></div>
      <input type="hidden" name="check_keywords" value="yes" />
      <input type="hidden" name="area" value="default" />
    </form>
</div>
<script type="text/javascript">$('#searchbox').show(0);</script>
    <p class="logo">
       <img class="logo" src="_static/logos.png" alt="Logos" usemap="#logosmap" />
    </p>

        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="right" >
          <a href="index.html" title="Adenine (A Data ExploratioN pIpeliNE)"
             >previous</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">ADENINE 0.1.4 documentation</a> &#187;</li> 
      </ul>
    </div>
    <div class="footer" role="contentinfo">
        &#169; Copyright 2016, Samuele Fiorini - Federico Tomasi - Annalisa Barla.
      Created using <a href="http://sphinx-doc.org/">Sphinx</a> 1.4.6.
    </div>
  </body>
</html>